weave/flow/playground_model.py

```diff
@@ -0,0 +1,220 @@
+                    f"LLM API returned an error: {response_payload['error']}"
+                )
+
+            # Assuming OpenAI-like structure: a list of choices, first choice has the message
+            output_message_dict = response_payload["choices"][0]["message"]
+
+            if self.return_type == "string":
+                return output_message_dict["content"]
+            elif self.return_type == "message":
+                return Message.model_validate(output_message_dict)
+            elif self.return_type == "json":
+                return json.loads(output_message_dict["content"])
+            else:
+                raise ValueError(f"Invalid return_type: {self.return_type}")
+        except (
+            KeyError,
+            IndexError,
+            TypeError,
+            AttributeError,
+            json.JSONDecodeError,
+        ) as e:
+            raise RuntimeError(
+                f"Failed to extract message from LLM response payload. Response: {api_response.response}, Error: {e}"
+            ) from e
+
+
+def _prepare_llm_messages(
+    template_messages: Optional[list[Message]],
+    user_input: Union[Message, list[Message], str],
+) -> list[dict[str, Any]]:
+    """
+    Prepares a list of message dictionaries for the LLM API from a message template and user input.
+    Helper function for PlaygroundModel.predict.
+    Returns a list of message dictionaries.
+    """
+    final_messages_dicts: list[dict[str, Any]] = []
+
+    # 1. Initialize messages from template
+    if template_messages:
+        for msg_template in template_messages:
+            msg_dict = msg_template.model_dump(exclude_none=True)
+            final_messages_dicts.append(msg_dict)
+
+    # 2. Parse user_input messages
+    user_input_list: list[Message]
+    if isinstance(user_input, str):
+        user_input_list = [
+            Message(
+                role="user",
+                content=user_input,
+                name=None,
+                function_call=None,
+                tool_call_id=None,
+            )
+        ]
+    elif isinstance(user_input, Message):
+        user_input_list = [user_input]
+    elif isinstance(user_input, list):
+        user_input_list = user_input  # Assumes it's List[Message]
+    else:
+        raise TypeError(
+            f"user_input must be a string, Message, or list of Messages, got {type(user_input)}"
+        )
+
+    # 3. Append user_input messages
+    for u_msg in user_input_list:
+        if not isinstance(u_msg, Message):
+            # This check is important if user_input_list could come from an untyped list
+            raise TypeError(
+                f"Each item in user_input list must be of type Message, got {type(u_msg)}"
+            )
+        final_messages_dicts.append(u_msg.model_dump(exclude_none=True))
+
+    return final_messages_dicts
+
+
+def parse_params_to_litellm_params(
+    params_source: Union[LLMStructuredCompletionModelDefaultParams, dict[str, Any]],
```

Comment:
There are a few places in this PR where you accept multiple input types and then do some sort of casting.  See `casting.py` which might provide a cleaner and more reusable way of doing this:
https://github.com/wandb/weave/blame/master/weave/flow/casting.py